_posts/2022-09-27-building-container-images.adoc

```diff
@@ -0,0 +1,173 @@
+---
+title: Building Containers in Rootless Mode on OpenShift
+layout: post
+author: Mario Loriedo
+description: >-
+  In this blog post, we are going to see how to configure Eclispe Che to run `podman build` from a remote development environment.
+categories: []
+keywords: ['buildah', 'podman']
+slug: /@mloriedo/building-container-images
+---
+
+image::/assets/img/building-container-images/che-inner-loop-optimized.png[Developer inner loop including a container build]
+{nbsp} +
+{nbsp} +
+Nowadays the typical developer flow, the inner-loop, includes quite often the build of a `Dockerfile`. Despite that Eclipse Che did not support building container images on OpenShift. That is because it involved granting permissive OpenShift privileges to Che users and we did not want to require that.
+
+The good news is that today a few options are available to build containers without compromising the underlying OpenShift cluster security. One of these is running in rootless mode and that is what we are going to discuss in this short blog post.
+
+Although it's now easy to link:https://github.com/containers/buildah/blob/main/docs/tutorials/05-openshift-rootless-build.md[run `buildah` or `podman build` in rootless mode], on OpenShift it requires the grant of some non-default link:https://man7.org/linux/man-pages/man7/capabilities.7.html[Linux capabilities]: `CAP_SETGID` and `CAP_SETUID`. It is possible to provide those capabilities to an Eclipse Che workspace following these steps:
+
+1. <<step_1>>
+2. <<step_2>>
+3. <<step_3>>
+4. <<step_4>>
+
+The first 3 steps setup the OpenShift cluster and Eclipse Che. These are administration tasks. The last step is for users that start a workspace: every workspace that requires `CAP_SETGID` and `CAP_SETUID` capability needs that Devfile attribute.
+
+== STEP 1: Creating an OpenShift Security Context Constraint [[step_1]]
+
+In OpenShift, permissions for Pods are controlled with link:https://docs.openshift.com/container-platform/latest/authentication/managing-security-context-constraints.html[security context constraints (SCC)]. OpenShift includes some pre-defined SCCs but the `restricted` SCC (the default one) does not provide enough capabilities and the `non-root` SCC provides more capabilities than required. To be able to build containers but avoid granting unrequired privileges we need to define an ad-hoc SCC that we call `container-build`. 
+
+Use an admin account to create it on an OpenShift cluster with the following command:
+
+[source,bash]
+----
+kubectl apply -f - <<EOF
+apiVersion: security.openshift.io/v1
+kind: SecurityContextConstraints
+metadata:
+  name: container-build
+allowHostDirVolumePlugin: false
+allowHostIPC: false
+allowHostNetwork: false
+allowHostPID: false
+allowHostPorts: false
+allowPrivilegeEscalation: true
+allowPrivilegedContainer: false
+allowedCapabilities:
```

Comment:
Indentation in the code block is missing

Reply:
Nevermind, I expected a `spec` block or similar; didn't realize it was all top level -- that's strange.